broker.go

```diff
@@ -30,17 +30,23 @@ import (
 //
 // A Node can be shared across multiple pipelines.
 type Broker struct {
-	nodes  map[NodeID]Node
+	nodes  map[NodeID]*nodeUsage
```

Comment:
This all looks good to me, but I wonder if the "right" thing to do would be to delete this map altogether, and rely on the garbage collector to do the job of keeping track of references. We don't do anything special when the nodes are deleted (like clean up), and the only purpose of this maps seems to be to keep track of mapping IDs to the actual nodes.

It's too bad Go doesn't have the idea of a weak map for this, which would do this basically automatically.

Removing this or changing it much would be a possibly breaking change though, so probably not for this PR :)

Reply:
Your comment got me to wondering... should we add an optional `Closer()` interface for Nodes, especially if we're going to start deleting them?

Reply:
I really appreciate both your comments on this, it's an interesting discussion, and I think it would definitely be worth exploring adding this... but maybe like @swenson said 'probably not for this PR' 😄 

We currently allow nodes to be 'swapped out' 😬 which leaves the old node in limbo anyway, so  nothing in this PR makes this a new problem.

Reply:
Deleting a node might create a new problem with any node that has state.  

For example, the [gated.Filter](https://github.com/hashicorp/go-eventlogger/blob/cca445805f24db4b4804eb58bb766ae94867b710/filters/gated/gated.go#L75) has events in memory which haven't been flushed and will periodically check to see if they've "expired" which flushes them (sending them along to the next node).  Providing an optional `Closer()` interface which is called before deleting nodes might allow these "stateful" nodes to be safely deleted.